### ISSUE TITLE
[Question] write accessor erase while traversal of concurrent_hash_map using iterators

### ISSUE THREAD

**Issue by yiakwy** [oneapi-src/oneTBB#183]
are the following codes safe with intel tbb::concurrent_hash_map:

```c++
// PointCloud is specialization of concurrent_hash_map template 
for (PointCloud::iterator it=pointCloud.begin(); it !=pointCloud.end();)
{
  auto snapshot = *it++;
  auto point = std::move(snapshot.second); // a shared pointer
  if (!point) {
    LOG_INFO(...);
    continue;
  }

  if (some condition fulfilled) {
    // try to delete the element
    PointCloud::accessor writer;
    if (!pointCloud.find(writer, point->key() ) ) {
      continue;
    }
    if (!pointCloud.erase(writer) ) {
      continue; // already erased by another thread
    }
  
   // acquire a lock for this element 

  //  do something here ...

   // the element is successfully erased by this call; note that the memory will not be released back to system but marked as free for next new elements.
   point.reset(); // release ownship of the resource
   writer.release(); // release the lock hold for this key

  } // end of condition fulfilled

} //  end of for

```

I have read many articles from intel tbb zone but none of them solved my question. And I want to know understand how this is implemented (the source code is unreadable for human beings)

Few other questions:
1) During debug I occasionally found that iterator pointed to the same resource just erased would appear immediately again (only once, I have to maintain a thread local `unorderd_map` for sanity check)
2) During multi threads, iterators pointed to the different resources with the same hash key are acceptable, I will erase it again and merge the results for this thread and dump the erased elements to disk later.

**Comment by yiakwy** [comment]
@kad @argretzi @szollin @szollin @rdower

**Comment by yiakwy** [comment]
@alexey-katranov

**Comment by anton-malakhov** [comment]
You cannot do it legally with concurrent_hash_map iterators as any operation with it invalidates iterators. Did you read https://software.intel.com/en-us/blogs/2010/05/14/traversing-concurrent_hash_map-concurrently ? It offers one soluntion with many restrictions but the code above does not confirm to them:

> Although it is not documented, you can traverse tbb::concurrent_hash_map concurrently if you are fine with the following restrictions and possible workarounds for them:
>1. You cannot traverse concurrently with an erase() operation. Use a mutex around erase and traversal code if you still need both running in different threads.
>2. Some items can be missed due to concurrent rehashing. Try to reserve enough buckets before traversal using rehash() to prevent this or repeat the pass.
>3. Some items can be visited twice or more times. Try to reserve enough buckets before traversal using rehash() to prevent this or put everything to a local serial map first to remove duplicates.
>4. Access through iterators is not protected. Know what are you doing, e.g. use additional synchronization primitives like tbb::atomic to protect one member or acquire accessor to protect the rest.

If you are interested in the details of the algorithms, please read https://arxiv.org/abs/1509.02235

**Comment by yiakwy** [comment]
thank @anton-malakhov for fast response:

## About articles

1. Thank your for the recommendation. I just read it. But honestly speaking, I haven't digested the data structure and operations very well. But I got some hints from `Per-bucket concurrent rehashing` method:

see [here](https://github.com/jckarter/tbb/blob/master/include/tbb/concurrent_hash_map.h), `class concurrent_hash_map : protected internal::hash_map_base`, we see that a concurrent_hash_map is a structure as described by the paper with

1)
```c++
       ...
        //  1. many buckets, and each bucket holds a light-weight lock designed for containers
        struct bucket : tbb::internal::no_copy {
            //! Mutex type for buckets
            typedef spin_rw_mutex mutex_t;
            //! Scoped lock type for mutex
            typedef mutex_t::scoped_lock scoped_t;
            mutex_t mutex;
            node_base *node_list;
        };
        ...
        //! Segment pointer
        typedef bucket *segment_ptr_t;
        //! Segment pointers table type
        typedef segment_ptr_t segments_table_t[pointers_per_table];
        ...
        //! Segment pointers table. Also prevents false sharing between my_mask and my_size
        segments_table_t my_table;

```

Simply put , array of linked list \(bucket_t\). 

2) Resizing

The number of buckets is the the number of slots hashed. The copy is lazy copy, i.e.: 

Using pow of 2 hashing,  an old bucket i (may contain 1 or 2 nodes if we have a good hasher) will be move immediately to a new slot in a new allocated array only when i_n =`HASHER(i) mod 2*capacity > capacity`, where` i = i_n mod capacity`.

Figure 1 & Figure 2 from the paper are ambiguous. 

It also suggests in `Reactive synchronization`, during iteration a key is not found if it was removed by a concurrent rehashing of a bucket. Hence I wrote `find` in the loop again.

```c++
    if (!pointCloud.find(writer, point->key() ) ) {
      continue;
    }
```

The proposed algorithm in insert method propose a recursive lookup method to avoid such cases:

```c++
51 label restart:
52   b = acquire_bucket( h & m );
53   p = b.search( key );
54   if( p == 0 ) {
55     m_now = my_mask.fetch();
56   if( (h & m)!=(h & m_now)
57   && is_race(h, m) ) {
58     unlock( b );
59     m = m_now;
60     goto restart;
61    }
```

Where my_mask is an atomic variable guarded by CPU execution sequences. I wonder whether this related to my cases.

## Bucket Lock

The interesting part is that a bucket is locked using rw_lock when it is acquired. How does intel tbb make sure that for different bucket b1, and b2, with limited mutexes, we lock them and allow them to work without waiting for each other?

```c++11
21 acquire_bucket( i ) {
22   // find bucket i
23   k = segment_index_of( i );
24   j = i - segment_base( k );
25   b = my_table[k][j];
26   lock( b );
27   if( is_marked_new( b ) )
28     rehash_bucket( b, i );
29   return b;
30 }
```
## About usage purposes

1. Mutex around concurrent_hash_map::erase() :
I don't want to do that, because this will lock the whole table. Hash Table find is nearly O(1) operation and then I can find it to acquire a write lock and erase it. That is the whole logics why I write in this manner. 

I personally believe that muext around concurrent_hash_map is meaningless.

2. I am fine with a chance of not visiting all your items in one pass

3. I agree that iteration over the table is not thread safe but thanks to these lines form then mentioned article:

>  It means that a thread doing concurrent traversal will not read garbage but on the other hand can read an old, not-yet-updated value. So despite the fact that the whole data-structure will be in a valid state at any given time, an iterator can miss some actually stored items or meet the same item twice

We want have segment fault error and actually I don't have in my experiments. I have maintain a thread local hash_table  (still O(1) for query) for sanity check. Actually I have double check.

**Comment by nofuturre** [comment]
@yiakwy is this issue still relevant?

**Comment by nofuturre** [comment]
If anyone encounter this issue in the future please open new issue with a link to this one
